### PR TITLE
Spack TPLs & GitLab-CI

### PR DESCRIPTION
**Pull request by mdavis36**
# Summary
This PR brings in gitlab-ci testing along with changes to the TPL management system for maintaining our TPL builds across various compilers and configurations.

- TPLs are to be managed through `spack` removing the need to maintain our own TPL build system for each of our dependencies.
- A set of dev-tool scripts are supplied to help Spheral developers with managing Spack for Spheral builds. These can be found at `scripts/dev-tools`
- Gitlab will run a suite of tests on LLNL CZ machines for PR/branches created by verified LLNL developers of the Spheral project. 
  - TOSS
    - gcc@8.3.1^mvpich2 
    - gcc@8.1.0^mvapich2 Debug
    - gcc@8.3.1~mpi
    - clang@9.0.0^mvapich2
  - BLUEOS
    - gcc@8.1.0^spectrum-mpi
    - gcc@8.1.0~mpi Debug  

## Other Changes
- Axom, Conduit and QHull now use shared libs.
- Fixes Clang Builds

## TODO
- [x] Documentation pass
- [x] Build mirrors in shared space for RZ and CZ users.
- [x] Tag release of Spheral before merging. That will be the l$t supported version of Spheral with the old TPL system. This PR will most likely break builds if a user tries to use the old TPL system.

### CODE REVIEW

**Comment by mdavis36** [comment]
LGTM

**Review by jmikeowen — DISMISSED** [review]
This is a big change for developers.  I'm sure we'll be iterating on the docs as questions come up.  Good work on getting both Spack and CI up and running!

**Review by MishaZakharchanka — DISMISSED** [review]

**Review by jmpearl — APPROVED** [review]

File: src/Mesh/Mesh.cc [hunk 1]
```diff
@@ -845,7 +845,7 @@ generateDomainInfo() {
 
   // Determine which process owns each shared node.  We use the convention that the process
   // with the lowest rank wins.
-  unordered_map<unsigned, unsigned> nodeOwner;
+  boost::unordered_map<unsigned, unsigned> nodeOwner;
```
Comment (ptsuji) [inline src/Mesh/Mesh.cc:848]: Is there a big reason for switching to boost::unordered_map here?

Comment (mdavis36) [reply src/Mesh/Mesh.cc:848]: I believe we have always been using the boost implementation here. `using namespace boost` & `using ::boost::unordered_map` is included at the top of this file. I think I added it for clarity on which `unordered_map` we are calling at the moment.

**Review by ptsuji — APPROVED** [review]

**Comment by mdavis36** [comment]
@ptsuji @jmikeowen @MishaZakharchanka @jmpearl @brbass Our tests are coming back green and and I believe I've addressed our talking points from the meeting last Friday. I've also tagged a release for [pre-spack Spheral](https://github.com/LLNL/spheral/releases/tag/2022.2.0-pre-spack). If there aren't any last requests or comments I will merge this

**Review by brbass — APPROVED** [review]
Looks great, @mdavis36!

**Pull request merged**